File: backend/app/schemas/vision.py
"""
AEGON Vision Intelligence — Pydantic Schemas
Enterprise-grade DTOs for the multi-angle asset inspection system.
"""
from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Optional
from uuid import UUID, uuid4

from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# Enums
# ---------------------------------------------------------------------------

class VisionViewType(str, Enum):
    TOP = "top"
    BOTTOM = "bottom"
    FRONT = "front"
    REAR = "rear"
    LEFT = "left"
    RIGHT = "right"
    COMPOSITE_360 = "360"


class DefectSeverity(str, Enum):
    CRITICAL = "critical"
    HIGH = "high"
    MEDIUM = "medium"
    LOW = "low"
    NONE = "none"


class InspectionStatus(str, Enum):
    IDLE = "idle"
    UPLOADING = "uploading"
    ANALYZING = "analyzing"
    COMPLETE = "complete"
    FAILED = "failed"


class MaintenancePriority(str, Enum):
    IMMEDIATE = "immediate"
    WITHIN_7_DAYS = "within_7_days"
    WITHIN_30_DAYS = "within_30_days"
    WITHIN_90_DAYS = "within_90_days"
    SCHEDULED = "scheduled"
    NONE_REQUIRED = "none_required"


# ---------------------------------------------------------------------------
# Core Finding DTO
# ---------------------------------------------------------------------------

class BoundingBoxMock(BaseModel):
    """Mock bounding box — ready for real YOLO/DETR integration."""
    x: float = Field(ge=0.0, le=1.0, description="Normalised x position (0-1)")
    y: float = Field(ge=0.0, le=1.0, description="Normalised y position (0-1)")
    width: float = Field(ge=0.0, le=1.0, description="Normalised width (0-1)")
    height: float = Field(ge=0.0, le=1.0, description="Normalised height (0-1)")
    label: str
    confidence: float = Field(ge=0.0, le=1.0)


class DefectFinding(BaseModel):
    """
    A single defect detected during vision inspection.
    Designed to be output-compatible with YOLOv11 / RT-DETR / SAM results.
    """
    finding_id: UUID = Field(default_factory=uuid4)
    defect_type: str
    severity: DefectSeverity
    confidence: float = Field(ge=0.0, le=1.0, description="AI confidence 0.0-1.0")
    description: str
    recommended_action: str
    bounding_box: Optional[BoundingBoxMock] = None
    area_affected_pct: float = Field(default=0.0, ge=0.0, le=100.0)


# ---------------------------------------------------------------------------
# Per-View Result
# ---------------------------------------------------------------------------

class ViewInspectionResult(BaseModel):
    """Result of a single-view AI inspection pass."""
    view_type: VisionViewType
    status: InspectionStatus
    image_filename: Optional[str] = None
    image_base64_preview: Optional[str] = None  # thumbnail only

    # AI output
    findings: list[DefectFinding] = Field(default_factory=list)
    view_health_score: float = Field(ge=0.0, le=100.0, default=100.0)
    defect_count: int = 0
    critical_count: int = 0
    processing_time_ms: float = 0.0
    inference_engine: str = "MockInferenceEngine"

    # Metadata
    inspected_at: Optional[datetime] = None
    operator: Optional[str] = None


# ---------------------------------------------------------------------------
# 360° Composite
# ---------------------------------------------------------------------------

class CompositeAnalysis(BaseModel):
    """Fused 360° analysis — synthesised from all view results."""
    overall_health_score: float = Field(ge=0.0, le=100.0)
    risk_score: float = Field(ge=0.0, le=100.0)
    quality_score: float = Field(ge=0.0, le=100.0)
    inspection_confidence: float = Field(ge=0.0, le=1.0)
    total_defects: int
    critical_defects: int
    views_inspected: int
    remaining_useful_life_years: float
    maintenance_priority: MaintenancePriority
    maintenance_cost_estimate_usd: float
    asset_readiness_pct: float
    deployment_status: str


# ---------------------------------------------------------------------------
# Executive Summary
# ---------------------------------------------------------------------------

class ExecutiveSummary(BaseModel):
    """AI-generated executive narrative."""
    summary_id: UUID = Field(default_factory=uuid4)
    narrative: str
    key_findings: list[str] = Field(default_factory=list)
    risk_matrix: dict[str, int] = Field(default_factory=dict)  # severity -> count
    recommendations: list[str] = Field(default_factory=list)
    generated_at: datetime = Field(default_factory=datetime.utcnow)


# ---------------------------------------------------------------------------
# Digital Twin State
# ---------------------------------------------------------------------------

class DigitalTwinState(BaseModel):
    """Real-time state of the asset Digital Twin, updated per inspection."""
    asset_name: str = "Asset Under Inspection"
    health_score: float = Field(ge=0.0, le=100.0, default=100.0)
    risk_score: float = Field(ge=0.0, le=100.0, default=0.0)
    temperature_celsius: float = 25.0
    maintenance_status: str = "Up to Date"
    inspection_progress_pct: float = Field(ge=0.0, le=100.0, default=0.0)
    views_completed: list[VisionViewType] = Field(default_factory=list)
    last_updated: datetime = Field(default_factory=datetime.utcnow)


# ---------------------------------------------------------------------------
# Inspection Session
# ---------------------------------------------------------------------------

class InspectionSession(BaseModel):
    """A complete inspection session encompassing all views."""
    session_id: UUID = Field(default_factory=uuid4)
    asset_id: Optional[str] = None
    asset_name: str = "Enterprise Asset"
    operator: str = "System"
    started_at: datetime = Field(default_factory=datetime.utcnow)
    completed_at: Optional[datetime] = None
    status: InspectionStatus = InspectionStatus.IDLE

    # Per-view results keyed by view type
    view_results: dict[str, ViewInspectionResult] = Field(default_factory=dict)

    # 360° analysis (populated after composite)
    composite: Optional[CompositeAnalysis] = None
    executive_summary: Optional[ExecutiveSummary] = None
    digital_twin: DigitalTwinState = Field(default_factory=DigitalTwinState)


# ---------------------------------------------------------------------------
# API Request / Response Schemas
# ---------------------------------------------------------------------------

class AnalyzeViewRequest(BaseModel):
    session_id: str
    view_type: VisionViewType
    asset_id: Optional[str] = None
    asset_name: Optional[str] = None
    operator: Optional[str] = "System"


class Generate360Request(BaseModel):
    session_id: str


class VisionStatistics(BaseModel):
    total_inspections: int
    avg_health_score: float
    avg_risk_score: float
    most_common_defect: str
    critical_defect_rate_pct: float
    avg_inspection_time_seconds: float

File: frontend/src/types/vision.ts
/**
 * AEGON Vision Intelligence — TypeScript Types
 * Mirrors the backend Pydantic schemas for strict end-to-end typing.
 */

export type VisionViewType = "top" | "bottom" | "front" | "rear" | "left" | "right" | "360";

export type DefectSeverity = "critical" | "high" | "medium" | "low" | "none";

export type InspectionStatus = "idle" | "uploading" | "analyzing" | "complete" | "failed";

export type MaintenancePriority =
  | "immediate"
  | "within_7_days"
  | "within_30_days"
  | "within_90_days"
  | "scheduled"
  | "none_required";

export interface BoundingBoxMock {
  x: number;
  y: number;
  width: number;
  height: number;
  label: string;
  confidence: number;
}

export interface DefectFinding {
  finding_id: string;
  defect_type: string;
  severity: DefectSeverity;
  confidence: number; // 0.0-1.0
  description: string;
  recommended_action: string;
  bounding_box?: BoundingBoxMock;
  area_affected_pct: number;
  heatmap_data?: string;
  mask_points: number[][];
  estimated_repair_cost: number;
  estimated_repair_time_mins: number;
}

export interface ViewInspectionResult {
  view_type: VisionViewType;
  status: InspectionStatus;
  image_filename?: string;
  findings: DefectFinding[];
  view_health_score: number;
  defect_count: number;
  critical_count: number;
  processing_time_ms: number;
  inference_engine: string;
  model_version: string;
  gpu_status: string;
  cpu_usage_pct: number;
  memory_usage_mb: number;
  queue_position: number;
  inspected_at?: string;
  operator?: string;
}

export interface CompositeAnalysis {
  overall_health_score: number;
  risk_score: number;
  quality_score: number;
  inspection_confidence: number;
  total_defects: number;
  critical_defects: number;
  views_inspected: number;
  remaining_useful_life_years: number;
  maintenance_priority: MaintenancePriority;
  maintenance_cost_estimate_usd: number;
  asset_readiness_pct: number;
  deployment_status: string;
}

export interface ExecutiveSummary {
  summary_id: string;
  narrative: string;
  key_findings: string[];
  risk_matrix: Record<string, number>;
  recommendations: string[];
  generated_at: string;
}

export interface DigitalTwinState {
  asset_name: string;
  health_score: number;
  risk_score: number;
  temperature_celsius: number;
  pressure_psi: number;
  rotation_rpm: number;
  maintenance_status: string;
  inspection_progress_pct: number;
  views_completed: VisionViewType[];
  historical_health_trend: number[];
  historical_risk_trend: number[];
  last_updated: string;
}

export interface InspectionSession {
  session_id: string;
  asset_id?: string;
  asset_name: string;
  operator: string;
  started_at: string;
  completed_at?: string;
  status: InspectionStatus;
  view_results: Record<string, ViewInspectionResult>;
  composite?: CompositeAnalysis;
  executive_summary?: ExecutiveSummary;
  digital_twin: DigitalTwinState;
}

export interface UploadViewResponse {
  session_id: string;
  view_result: ViewInspectionResult;
  digital_twin?: DigitalTwinState;
}

export interface Composite360Response {
  session_id: string;
  composite: CompositeAnalysis;
  executive_summary: ExecutiveSummary;
  digital_twin?: DigitalTwinState;
}

export interface InspectionHistoryItem {
  session_id: string;
  asset_name: string;
  operator: string;
  started_at: string;
  completed_at?: string;
  views_inspected: number;
  total_defects: number;
  critical_defects: number;
  health_score: number;
  risk_score: number;
  maintenance_priority: MaintenancePriority;
  status: InspectionStatus;
}

export interface VisionStatistics {
  total_inspections: number;
  avg_health_score: number;
  avg_risk_score: number;
  most_common_defect: string;
  critical_defect_rate_pct: number;
  avg_inspection_time_seconds: number;
  avg_queue_time_ms: number;
  active_gpu_utilization_pct: number;
}

export interface CreateMaintenanceTicketRequest {
  session_id: string;
  asset_id?: string;
  finding_id: string;
  defect_type: string;
  priority: MaintenancePriority;
  estimated_cost: number;
}

// UI-only types
export interface ViewPanelState {
  viewType: VisionViewType;
  label: string;
  status: InspectionStatus;
  previewUrl?: string;
  result?: ViewInspectionResult;
}
